custom_components/bosch_alarm/__init__.py

```diff
@@ -30,38 +33,27 @@ async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
             host=entry.data[CONF_HOST], port=entry.data[CONF_PORT],
             automation_code=entry.data.get(CONF_PASSWORD, None),
             installer_or_user_code=entry.data.get(CONF_INSTALLER_CODE, entry.data.get(CONF_USER_CODE, None)))
-    try:
-        await panel.connect()
-    except asyncio.exceptions.TimeoutError:
-        _LOGGER.warning("Initial panel connection timed out...")
-    except:
-        logging.exception("Initial panel connection failed")
+
+    # The config flow sets the entries unique id to the serial number if available
+    # If the panel doesn't expose it's serial number, use the entry id as a unique id instead.
+    unique_id = entry.unique_id or entry.entry_id
+
+    data = PanelConnection(panel, unique_id, entry.data[CONF_MODEL])
 
     hass.data.setdefault(DOMAIN, {})
-    hass.data[DOMAIN][entry.entry_id] = panel
-
-    def setup():
-        # Some panels don't support retrieving a serial number.
-        # We still need some form of identifier, so fall back
-        # to the entry id.
-        if not panel.serial_number:
-            panel.serial_number = entry.entry_id
-
-        # Remove old devices using the panel model as an identifier
-        dr = device_registry.async_get(hass)
-        for device_entry in device_registry.async_entries_for_config_entry(dr, entry.entry_id):
-            if (DOMAIN, panel.model) in device_entry.identifiers:
-                dr.async_remove_device(device_entry.id)
-
-        hass.async_create_task(
-            hass.config_entries.async_forward_entry_setups(entry, PLATFORMS))
-    if panel.connection_status():
-        setup()
-    else:
-        panel.connection_status_observer.attach(
-                lambda: panel.connection_status() and setup())
+    hass.data[DOMAIN][entry.entry_id] = data
+
+    # Remove old devices using the panel model as an identifier
```

Comment:
Since we're already doing config entry update, would it work to stick this there as well? This way we won't have to keep running this on every startup